### PR TITLE
Feature/101 edit challenge page

### PR DESCRIPTION
**Pull request by SashaBondarchuk**
https://trello.com/c/1SFuG2eC/101-edit-challenge-page

https://github.com/BinaryStudioAcademy/bsa-2023-leetwars/assets/56970649/131b6d0f-63e0-412a-808b-bbef286360a7

Additionally I have fixed 2 followng bugs (see images in production below)

1). [inappropriate size of challenge block](https://trello.com/c/P5VSQg1w/176-bug64-main-page-the-challenges-block-is-narrow-when)
image
2). star position is absolute
image

### CODE REVIEW

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/ChallengeService.cs [hunk 1]
```diff
@@ -175,14 +177,73 @@ public async Task<ChallengeFullDto> CreateChallengeAsync(NewChallengeDto challen
...
+            if (currentUser.Id != challengeEditDto.CreatedBy)
+            {
+                throw new InvalidOperationException("The user cannot modify this challenge");
+            }
+
+            var challenge = await GetChallengeByIdAsync(challengeEditDto.Id);
+            if (challenge is null)
+            {
+                throw new NotFoundException(nameof(Challenge));
+            }
```
Comment (Limbo2332) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/ChallengeService.cs]:
```suggestion
            var challenge = await GetChallengeByIdAsync(challengeEditDto.Id)
                ?? throw new NotFoundException(nameof(Challenge));
```

File: backend/LeetWars.Core/LeetWars.Core.Common/DTO/Challenge/ChallengeEditDto.cs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+﻿using LeetWars.Core.Common.DTO.ChallengeVersion;
+using LeetWars.Core.Common.DTO.Tag;
+using LeetWars.Core.DAL.Entities;
+using LeetWars.Core.DAL.Enums;
+
+namespace LeetWars.Core.Common.DTO.Challenge
+{
+    public class ChallengeEditDto : Entity<long>
+    {
+        public long CreatedBy { get; set; }
```
Comment (Limbo2332) [inline backend/LeetWars.Core/LeetWars.Core.Common/DTO/Challenge/ChallengeEditDto.cs]: Why not AuditEntity then?

File: backend/LeetWars.Core/LeetWars.Core.Common/DTO/Challenge/ChallengePreviewDto.cs [hunk 3]
```diff
@@ -7,24 +7,16 @@ namespace LeetWars.Core.Common.DTO.Challenge
 {
     public class ChallengePreviewDto : Entity<long>
     {
+        public long CreatedBy { get; set; }
```
Comment (Limbo2332) [inline backend/LeetWars.Core/LeetWars.Core.Common/DTO/Challenge/ChallengePreviewDto.cs]: see previous comment

File: frontend/src/app/modules/challenges/challenge-creation/challenge-creation.component.ts [hunk 4]
```diff
@@ -73,41 +75,71 @@ export class ChallengeCreationComponent extends BaseComponent implements OnInit
...
     }
 
     ngOnInit(): void {
-        this.getLanguages();
-        this.getTags();
-        this.getChallengeLevels();
+        this.activatedRoute.paramMap.subscribe((params: ParamMap) => {
+            const challengeId = params.get('id')!;
+
+            this.isEditMode = challengeId !== null;
```
Comment (Limbo2332) [inline frontend/src/app/modules/challenges/challenge-creation/challenge-creation.component.ts]:
```suggestion
            const challengeId = +params.get('id');

            this.isEditMode = !isNaN(challengeId);
```

and parsing in 98 line is unnecessary

File: frontend/src/app/modules/main/challenge/challenge.component.ts [hunk 5]
```diff
@@ -26,8 +26,14 @@ export class ChallengeComponent {
 
     public getLanguageIconUrl = getLanguageIconUrl;
 
+    public canEdit: boolean;
+
     private user: IUser;
 
+    ngOnInit(): void {
+        this.canEdit = this.challenge.createdBy === this.user.id;
+    }
```
Comment (Limbo2332) [inline frontend/src/app/modules/main/challenge/challenge.component.ts]: move into constructor subscribe

File: frontend/src/app/shared/models/challenge/challenge-preview.ts [hunk 6]
```diff
@@ -6,6 +6,7 @@ import { Entity } from '../common/entity';
 
 export interface IChallengePreview extends Entity<number> {
     authorName: string;
+    createdBy: number;
```
Comment (Limbo2332) [inline frontend/src/app/shared/models/challenge/challenge-preview.ts]: the same as backend

**Review by Limbo2332 — CHANGES_REQUESTED** [review]

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/ChallengeService.cs [hunk 7]
```diff
@@ -175,14 +177,73 @@ public async Task<ChallengeFullDto> CreateChallengeAsync(NewChallengeDto challen
...
             return await GetChallengeFullDtoByIdAsync(challenge.Id);
         }
 
+        private void UpdateChallengeProperties(Challenge challenge, ChallengeEditDto challengeEditDto)
+        {
+            challenge.Title = challengeEditDto.Title;
+            challenge.LevelId = challengeEditDto.LevelId;
+            challenge.Instructions = challengeEditDto.Instructions;
+            challenge.Category = challengeEditDto.Category;
+        }
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/ChallengeService.cs]: cannot you use AutoMapper for this?

File: frontend/src/app/modules/challenges/challenge-creation/challenge-creation.utils.ts [hunk 8]
```diff
@@ -114,11 +119,49 @@ export function prepareChallengeDto(challenge: INewChallenge): INewChallenge {
...
+            id: version.id,
+            createdAt: version.createdAt.toString(),
+            languageId: version.languageId,
+            initialSolution: version.initialSolution,
+            completeSolution: version.completeSolution,
+            preloadedCode: version.preloadedCode,
+            testCases: version.testCases,
+            exampleTestCases: version.exampleTestCases,
+        })),
+    };
```
Comment (tatianahutii) [inline frontend/src/app/modules/challenges/challenge-creation/challenge-creation.utils.ts]: maybe spread operator?

**Review by tatianahutii — CHANGES_REQUESTED** [review]

**Review by Limbo2332 — APPROVED** [review]

**Review by tatianahutii — APPROVED** [review]
LGTM

**Review by Den313 — APPROVED** [review]

**Pull request merged**